### PR TITLE
[ready to review] local feature detection rebased

### PR DESCRIPTION
**Pull request by ducha-aiki**
ToDo:

 - [x] Numerical tests for orientaton, affine_shape and fully integrated pipeline
 - [x] Make sure thas integrated pipeline works with batch

Breaking changes: CornerHarris now has not normalization and nms inside

### CODE REVIEW

File: kornia/__init__.py [hunk 1]
```diff
@@ -26,13 +26,12 @@
     adjust_brightness,
 )
 from kornia.contrib import (
-    spatial_soft_argmax2d,
```
Comment (edgarriba) [inline kornia/__init__.py]: why are you removing this ?

File: kornia/contrib/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from .spatial_soft_argmax2d import SpatialSoftArgmax2d, spatial_soft_argmax2d
+from kornia.geometry.spatial_soft_argmax import SpatialSoftArgmax2d, spatial_soft_argmax2d
```
Comment (edgarriba) [inline kornia/contrib/__init__.py]: why not relative path here ? at some point for the next release we should review the way we import/expose the api.

File: kornia/feature/__init__.py [hunk 3]
```diff
@@ -23,5 +39,13 @@
...
-    "SIFTDescriptor"
+    "scale_laf",
+    "SIFTDescriptor",
+    "PassLAF",
+    "ScaleSpaceDetector",
+    "LAFAffineShapeEstimator",
+    "PatchAffineShapeEstimator",
+    "LAFOrienter",
+    "PatchDominantGradientOrientation",
+    "raise_error_if_laf_is_not_valid"
```
Comment (edgarriba) [inline kornia/feature/__init__.py]: do we want to expose this or it's just for internal use ?

File: kornia/feature/affine_shape.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+class PatchAffineShapeEstimator(nn.Module):
+    def __init__(self, patch_size: int = 19, eps: float = 1e-10):
+        """Module, which estimates the second moment matrix of the patch gradients in order to determine the
+        affine shape of the local feature as in
+        A. Baumberg (2000). "Reliable feature matching across widely separated views".
+        In Proc.of CVPR. pp. I:1774--1781.
+        Args:
+            patch_size: int, default = 19
+            eps: float, for safe division, default is 1e-10"""
```
Comment (edgarriba) [inline kornia/feature/affine_shape.py]: the docstrings diverges from the others. I'm fine since you checked the docs are well generated.

File: kornia/feature/affine_shape.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                                raise_error_if_laf_is_not_valid,
+                                scale_laf, make_upright)
+from kornia.feature import extract_patches_from_pyramid
+
+
+class PatchAffineShapeEstimator(nn.Module):
+    def __init__(self, patch_size: int = 19, eps: float = 1e-10):
+        """Module, which estimates the second moment matrix of the patch gradients in order to determine the
+        affine shape of the local feature as in
+        A. Baumberg (2000). "Reliable feature matching across widely separated views".
```
Comment (edgarriba) [inline kornia/feature/affine_shape.py]: I think there's a way to generate a reference like in latex

File: kornia/feature/affine_shape.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        Args:
+            patch_size: int, default = 19
+            eps: float, for safe division, default is 1e-10"""
+        super(PatchAffineShapeEstimator, self).__init__()
+        self.patch_size: int = patch_size
+        self.gradient: nn.Module = SpatialGradient('sobel', 1)
+        self.eps: float = eps
+        sigma: float = float(self.patch_size) / math.sqrt(2.0)
+        self.weighting: torch.Tensor = get_gaussian_kernel2d((self.patch_size, self.patch_size), (sigma, sigma), True)
+        return
```
Comment (edgarriba) [inline kornia/feature/affine_shape.py]: is return really needed? :thinking:

File: kornia/feature/affine_shape.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        """Module, which estimates the second moment matrix of the patch gradients in order to determine the
+        affine shape of the local feature as in
+        A. Baumberg (2000). "Reliable feature matching across widely separated views".
+        In Proc.of CVPR. pp. I:1774--1781.
+        Args:
+            patch_size: int, default = 19
+            eps: float, for safe division, default is 1e-10"""
+        super(PatchAffineShapeEstimator, self).__init__()
+        self.patch_size: int = patch_size
+        self.gradient: nn.Module = SpatialGradient('sobel', 1)
```
Comment (edgarriba) [inline kornia/feature/affine_shape.py]: could you set the second argument explicit to make the code more readable ?

File: kornia/feature/affine_shape.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        self.eps: float = eps
+        sigma: float = float(self.patch_size) / math.sqrt(2.0)
+        self.weighting: torch.Tensor = get_gaussian_kernel2d((self.patch_size, self.patch_size), (sigma, sigma), True)
```
Comment (edgarriba) [inline kornia/feature/affine_shape.py]: would make explicit the last argument

File: kornia/feature/laf.py [hunk 9]
```diff
@@ -291,71 +307,76 @@ def generate_patch_grid_from_normalized_LAF(img: torch.Tensor,
...
+                           PS: int = 32,
+                           normalize_lafs_before_extraction: bool = True) -> torch.Tensor:
     """
     Extract patches defined by LAFs from image tensor.
     No smoothing applied, huge aliasing (better use extract_patches_from_pyramid)
     Args:
-        LAF: (torch.Tensor).
-        images: (torch.Tensor) images, LAFs are detected in
+        img: (torch.Tensor) images, LAFs are detected in
+        laf: (torch.Tensor).
```
Comment (edgarriba) [inline kornia/feature/laf.py]: add some description

File: kornia/feature/laf.py [hunk 10]
```diff
@@ -291,71 +307,76 @@ def generate_patch_grid_from_normalized_LAF(img: torch.Tensor,
...
         PS: (int) patch size, default = 32
-
+        normalize_lafs_before_extraction (bool) - if True, lafs are normalized to image size, default = True
```
Comment (edgarriba) [inline kornia/feature/laf.py]: the arguments that are optional I would start stating it: `normalize_lafs_before_extraction (bool, optional)`

File: kornia/feature/orientation.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                 patch_size: int = 32,
+                 num_angular_bins: int = 36, eps: float = 1e-8):
+        super(PatchDominantGradientOrientation, self).__init__()
+        self.patch_size = patch_size
+        self.num_ang_bins = num_angular_bins
+        self.gradient = SpatialGradient('sobel', 1)
+        self.eps = eps
+        self.angular_smooth = nn.Conv1d(1, 1, kernel_size=3, padding=2, bias=False, padding_mode="circular")
+        with torch.no_grad():
+            self.angular_smooth.weight[:] = torch.tensor([[[0.33, 0.34, 0.33]]])
```
Comment (edgarriba) [inline kornia/feature/orientation.py]: wouldn't be more clean with F.conv1d and just setting the kernel here ?

File: kornia/feature/responses.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+import torch
+import torch.nn as nn
+
+from kornia.filters import spatial_gradient, gaussian_blur2d
+
+
+def harris_response(input: torch.Tensor,
+                    k: Union[torch.Tensor, float] = 0.04,
+                    grads_mode: str = 'sobel',
+                    sigmas: Union[torch.Tensor, None] = None) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/feature/responses.py]: recently learnt that in this case we should use Optional. `Optional[torch.Tensor] = None`

File: kornia/geometry/conversions.py [hunk 13]
```diff
@@ -651,6 +653,71 @@ def denormalize_pixel_coordinates(
         torch.tensor(width), torch.tensor(height)
     ]).to(pixel_coordinates.device).to(pixel_coordinates.dtype)
 
-    factor: torch.Tensor = torch.tensor(2.) / (hw - 1)
+    factor: torch.Tensor = torch.tensor(2.) / (hw - 1).clamp(1e-8)
```
Comment (edgarriba) [inline kornia/geometry/conversions.py]: should we let the user  to set this epsilon value ?

File: kornia/utils/grid.py [hunk 14]
```diff
@@ -37,3 +37,36 @@ def create_meshgrid(
     base_grid: torch.Tensor = torch.stack(
         torch.meshgrid([xs, ys])).transpose(1, 2)  # 2xHxW
     return torch.unsqueeze(base_grid, dim=0).permute(0, 2, 3, 1)  # 1xHxWx2
+
+
+def create_meshgrid3d(
+        depth: int,
+        height: int,
+        width: int,
+        normalized_coordinates: Optional[bool] = True) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/utils/grid.py]: this should be `normalized_coordinates: bool = True`. I think Optional is for None values.

**Comment by edgarriba** [comment]
@ducha-aiki one more thing, please add here or in a next PR the new methods to docs

File: kornia/__init__.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/__init__.py]: Because I moved moved to kornia/geometry to be together with convsoftargmax, as we discussed before 
https://github.com/ducha-aiki/kornia/blob/rebased-local-feature-detection/kornia/geometry/spatial_soft_argmax.py#L409

File: kornia/contrib/__init__.py (hunk 2 above)
Comment (ducha-aiki) [reply kornia/contrib/__init__.py]: Good question :) I have moved the function with PyCharm "refactor" tool, everything related was moved automatically. btw, I definitely can recommend PyCharm as an IDE

File: kornia/feature/__init__.py (hunk 3 above)
Comment (ducha-aiki) [reply kornia/feature/__init__.py]: I think, it is useful enough, if users would like to do their own checks. But don`t have a strong opinion on this

File: kornia/feature/affine_shape.py (hunk 5 above)
Comment (ducha-aiki) [reply kornia/feature/affine_shape.py]: Thanks, I`ll check it and modify

File: kornia/feature/orientation.py (hunk 11 above)
Comment (ducha-aiki) [reply kornia/feature/orientation.py]: True.

File: kornia/feature/laf.py (hunk 10 above)
Comment (ducha-aiki) [reply kornia/feature/laf.py]: Disagree here, agree in other place. It is more like default value that optional - I don`t want to accept "None" here.

File: kornia/feature/__init__.py (hunk 3 above)
Comment (edgarriba) [reply kornia/feature/__init__.py]: maybe yes, but I would add it to the `check_` functions we were discussing in another thread

File: kornia/feature/laf.py (hunk 10 above)
Comment (edgarriba) [reply kornia/feature/laf.py]: would be great to somehow mark in the docs the parameters that are not strictly required

File: kornia/feature/affine_shape.py (hunk 4 above)
Comment (ducha-aiki) [reply kornia/feature/affine_shape.py]: This works, if add to rst:
.. autoclass:: PatchAffineShapeEstimator
   :members: forward

So 2 methods will be documented. I am doing it where forward() signature is different from standard single input-single output
